tbb4py: "python -m tbb test" sporadically fails ("Run OSX build" in conda-forge)

Build on tbb4py packages sometimes fails with the following error:
```
+ python -m tbb test
oneTBB version is 2021.2
oneTBB interface version is 12020
Traceback (most recent call last):
  File "/Users/runner/miniforge3/conda-bld/tbb_1617698700579/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/Users/runner/miniforge3/conda-bld/tbb_1617698700579/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/runner/miniforge3/conda-bld/tbb_1617698700579/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/python3.7/site-packages/tbb/__main__.py", line 20, in <module>
    exit(_main())
  File "/Users/runner/miniforge3/conda-bld/tbb_1617698700579/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/python3.7/site-packages/tbb/__init__.py", line 317, in _main
    return globals()['_' + args.name](*args.args)
  File "/Users/runner/miniforge3/conda-bld/tbb_1617698700579/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/python3.7/site-packages/tbb/__init__.py", line 55, in _test
    test(arg)
  File "/Users/runner/miniforge3/conda-bld/tbb_1617698700579/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/python3.7/site-packages/tbb/test.py", line 145, in test
    assert False, "Expected exception !"
AssertionError: Expected exception !
/Users/runner/miniforge3/conda-bld/tbb_1617698700579/test_tmp/run_test.sh: line 9: 11806 Segmentation fault: 11  python -m tbb test
```

For example:

- https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=300053&view=logs&j=58ac6aab-c4bc-5de2-2894-98e408cc8ec9&t=933f325c-924e-533d-4d95-e93b5843ce8b
- https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=292395&view=logs&j=5919e0e2-8634-58c1-7e2d-cc5398f52f3f&t=f395cd69-c266-53d9-bd41-0b417b09501f

Here is example of successful build: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=285277&view=logs&j=5919e0e2-8634-58c1-7e2d-cc5398f52f3f&t=f395cd69-c266-53d9-bd41-0b417b09501f

There were no affecting changes between successful and failed builds, so looks like the problem is sporadic.

--- a/python/tbb/test.py
+++ b/python/tbb/test.py
@@ -1,6 +1,6 @@
 #!/usr/bin/env python3
 #
-# Copyright (c) 2016-2021 Intel Corporation
+# Copyright (c) 2016-2022 Intel Corporation
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
 # you may not use this file except in compliance with the License.

--- a/python/tbb/test.py
+++ b/python/tbb/test.py
@@ -86,8 +86,10 @@ def test(arg=None):
     spin_flag = True
     def timeout_work(param):
         nonlocal spin_flag
+        say("[%d] Spin wait work start..." % get_tid())
         while spin_flag:
             time.sleep(0.0001) # yield equivalent
+        say("[%d] Work done." % get_tid())
         return str(param) if param != None else None
 
     ### Test copy/pasted from multiprocessing

--- a/python/tbb/test.py
+++ b/python/tbb/test.py
@@ -103,7 +103,7 @@ def test(arg=None):
             say("Good. Got expected timeout exception.")
         else:
             assert False, "Expected exception !"
-        spin_flag = False # unlock threads
+        spin_flag = False # unlock threads in timeout_work
 
     ### Test copy/pasted from multiprocessing
     pool = Pool(4)  # start worker threads

--- a/python/tbb/test.py
+++ b/python/tbb/test.py
@@ -82,6 +82,15 @@ def test(arg=None):
         say("[%d] Work done (%fms)." % (get_tid(), mseconds*10))
         return res
 
+    # special flag to to be set by thread calling async work
+    spin_flag = True
+    def timeout_work(param):
+        nonlocal spin_flag
+        while spin_flag:
+            # print(spin_flag)
+            time.sleep(0.0001) # yield equivalent
+        return str(param) if param != None else None
+
     ### Test copy/pasted from multiprocessing
     pool = Pool(4)  # start worker threads
 
@@ -105,11 +114,12 @@ def test(arg=None):
     assert next(it) == 4
 
     # Test apply_sync exceptions
-    result = pool.apply_async(time.sleep, (3,))
+    result = pool.apply_async(timeout_work, (None,))
     try:
         say(result.get(timeout=1))  # raises `TimeoutError`
     except TimeoutError:
         say("Good. Got expected timeout exception.")
+        spin_flag = False
     else:
         assert False, "Expected exception !"
     assert result.get() is None  # sleep() returns None
@@ -118,29 +128,33 @@ def test(arg=None):
         say("Result ready: %s" % s)
 
     # Test imap()
-    assert list(pool.imap(work, range(10, 3, -1), chunksize=4)) == list(map(
+    assert list(pool.imap(timeout_work, range(10, 3, -1), chunksize=4)) == list(map(
         str, range(10, 3, -1)))
 
     # Test imap_unordered()
-    assert sorted(pool.imap_unordered(work, range(10, 3, -1))) == sorted(map(
+    assert sorted(pool.imap_unordered(timeout_work, range(10, 3, -1))) == sorted(map(
         str, range(10, 3, -1)))
 
     # Test map_async()
-    result = pool.map_async(work, range(10), callback=cb)
+    spin_flag = True
+    result = pool.map_async(timeout_work, range(10), callback=cb)
     try:
         result.get(timeout=0.01)  # raises `TimeoutError`
     except TimeoutError:
         say("Good. Got expected timeout exception.")
+        spin_flag = False
     else:
         assert False, "Expected exception !"
     say(result.get())
 
     # Test imap_async()
-    result = pool.imap_async(work, range(3, 10), callback=cb)
+    spin_flag = True
+    result = pool.imap_async(timeout_work, range(3, 10), callback=cb)
     try:
         result.get(timeout=0.01)  # raises `TimeoutError`
     except TimeoutError:
         say("Good. Got expected timeout exception.")
+        spin_flag = False
     else:
         assert False, "Expected exception !"
     for i in result.get():
@@ -150,11 +164,13 @@ def test(arg=None):
         say("Item2:", i)
 
     # Test imap_unordered_async()
-    result = pool.imap_unordered_async(work, range(10, 3, -1), callback=cb)
+    spin_flag = True
+    result = pool.imap_unordered_async(timeout_work, range(10, 3, -1), callback=cb)
     try:
         say(result.get(timeout=0.01))  # raises `TimeoutError`
     except TimeoutError:
         say("Good. Got expected timeout exception.")
+        spin_flag = False
     else:
         assert False, "Expected exception !"
     for i in result.get():

--- a/python/tbb/test.py
+++ b/python/tbb/test.py
@@ -87,7 +87,6 @@ def test(arg=None):
     def timeout_work(param):
         nonlocal spin_flag
         while spin_flag:
-            # print(spin_flag)
             time.sleep(0.0001) # yield equivalent
         return str(param) if param != None else None
 

--- a/python/tbb/test.py
+++ b/python/tbb/test.py
@@ -103,7 +103,7 @@ def test(arg=None):
             say("Good. Got expected timeout exception.")
         else:
             assert False, "Expected exception !"
-        spin_flag = False
+        spin_flag = False # unlock threads
 
     ### Test copy/pasted from multiprocessing
     pool = Pool(4)  # start worker threads

--- a/python/tbb/test.py
+++ b/python/tbb/test.py
@@ -83,14 +83,27 @@ def test(arg=None):
         return res
 
     # special flag to to be set by thread calling async work
-    spin_flag = True
+    spin_flag = None
     def timeout_work(param):
-        nonlocal spin_flag
         say("[%d] Spin wait work start..." % get_tid())
         while spin_flag:
             time.sleep(0.0001) # yield equivalent
         say("[%d] Work done." % get_tid())
         return str(param) if param != None else None
+
+    def prepare_timeout_exception():
+        nonlocal spin_flag
+        spin_flag = True # lock threads in timeout_work
+
+    def check_timeout_exception(pool_object, func):
+        nonlocal spin_flag
+        try:
+            func(pool_object)
+        except TimeoutError:
+            say("Good. Got expected timeout exception.")
+        else:
+            assert False, "Expected exception !"
+        spin_flag = False
 
     ### Test copy/pasted from multiprocessing
     pool = Pool(4)  # start worker threads
@@ -115,49 +128,32 @@ def test(arg=None):
     assert next(it) == 4
 
     # Test apply_sync exceptions
+    prepare_timeout_exception()
     result = pool.apply_async(timeout_work, (None,))
-    try:
-        say(result.get(timeout=1))  # raises `TimeoutError`
-    except TimeoutError:
-        say("Good. Got expected timeout exception.")
-        spin_flag = False
-    else:
-        assert False, "Expected exception !"
+    check_timeout_exception(result, lambda result : say(result.get(timeout=1)))
     assert result.get() is None  # sleep() returns None
 
     def cb(s):
         say("Result ready: %s" % s)
 
     # Test imap()
-    assert list(pool.imap(timeout_work, range(10, 3, -1), chunksize=4)) == list(map(
+    assert list(pool.imap(work, range(10, 3, -1), chunksize=4)) == list(map(
         str, range(10, 3, -1)))
 
     # Test imap_unordered()
-    assert sorted(pool.imap_unordered(timeout_work, range(10, 3, -1))) == sorted(map(
+    assert sorted(pool.imap_unordered(work, range(10, 3, -1))) == sorted(map(
         str, range(10, 3, -1)))
 
     # Test map_async()
-    spin_flag = True
+    prepare_timeout_exception()
     result = pool.map_async(timeout_work, range(10), callback=cb)
-    try:
-        result.get(timeout=0.01)  # raises `TimeoutError`
-    except TimeoutError:
-        say("Good. Got expected timeout exception.")
-        spin_flag = False
-    else:
-        assert False, "Expected exception !"
+    check_timeout_exception(result, lambda result : result.get(timeout=0.01))
     say(result.get())
 
     # Test imap_async()
-    spin_flag = True
+    prepare_timeout_exception()
     result = pool.imap_async(timeout_work, range(3, 10), callback=cb)
-    try:
-        result.get(timeout=0.01)  # raises `TimeoutError`
-    except TimeoutError:
-        say("Good. Got expected timeout exception.")
-        spin_flag = False
-    else:
-        assert False, "Expected exception !"
+    check_timeout_exception(result, lambda result : result.get(timeout=0.01))
     for i in result.get():
         say("Item:", i)
     say("### Loop again:")
@@ -165,15 +161,9 @@ def test(arg=None):
         say("Item2:", i)
 
     # Test imap_unordered_async()
-    spin_flag = True
+    prepare_timeout_exception()
     result = pool.imap_unordered_async(timeout_work, range(10, 3, -1), callback=cb)
-    try:
-        say(result.get(timeout=0.01))  # raises `TimeoutError`
-    except TimeoutError:
-        say("Good. Got expected timeout exception.")
-        spin_flag = False
-    else:
-        assert False, "Expected exception !"
+    check_timeout_exception(result, lambda result : result.get(timeout=0.01))
     for i in result.get():
         say("Item1:", i)
     for i in result.get():

